### PR TITLE
feat: add ability to delimit records within one COS object

### PR DESCRIPTION
**Pull request by dpittner**
This PR implements: https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/issues/4 by allowing to specify an optional delimiter for records that are to be added into the same cos object.

Note that I hadn't had the chance to tests this with an actual Kafka.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,5 +2,10 @@
 .gradle/
 .project
 .settings/
+.idea/
+gradle/
+gradlew
+gradlew.bat
 bin/
 build/
+*.iml
```
Comment (mimaison) [inline .gitignore]: nit: can we add a newline at the end of the file?

File: src/main/java/com/ibm/eventstreams/connect/cossink/COSSinkConnectorConfig.java [hunk 2]
```diff
@@ -64,10 +64,16 @@
...
 
     static final String CONFIG_NAME_COS_OBJECT_RECORDS = "cos.object.records";
+
     private static final String CONFIG_DOCUMENTATION_COS_OBJECT_RECORDS =
             "The maximum number of Kafka records to group together into a single object storage object.";
     private static final String CONFIG_DISPLAY_COS_OBJECT_RECORDS = "Records per object";
 
+    static final String CONFIG_NAME_COS_OBJECT_RECORD_DELIMITER = "cos.object.record.delimiter";
+    private static final String CONFIG_DOCUMENTATION_COS_OBJECT_RECORD_DELIMITER =
+            "The delimited to separate records within a single object storage object.";
```
Comment (mimaison) [inline src/main/java/com/ibm/eventstreams/connect/cossink/COSSinkConnectorConfig.java]: typo: the "delimiter"

File: src/main/java/com/ibm/eventstreams/connect/cossink/partitionwriter/COSObject.java [hunk 3]
```diff
@@ -41,8 +41,15 @@
 
     private final List<SinkRecord> records = new LinkedList<>();
     private Long lastOffset;
-
-    COSObject() {
+    private final byte[] recordSeparatorBytes;
+
+    COSObject(String recordDelimiter) {
+        if (recordDelimiter != null) {
+            LOG.trace("> setting recordDelimiter to: {} of length", recordDelimiter, recordDelimiter.length());
```
Comment (mimaison) [inline src/main/java/com/ibm/eventstreams/connect/cossink/partitionwriter/COSObject.java]: It's missing the 2nd `{}` after length

**Comment by dpittner** [comment]
@mimaison comments should be addressed

**Comment by mimaison** [comment]
@dpittner Thanks for the quick update! The change looks good.

Did you get a chance to test it? Or can @lijs-cn validate this works as expected?

**Comment by mimaison** [comment]
@dpittner Can you also add the new configuration to the README?

**Comment by lijs-cn** [comment]
Hi @dpittner ,

Thanks for providing the fix, I've verified the PR, and found when I specify `,` as delimiter set in `cos.object.record.delimiter`, it works. 

But when I want to set newline as delimiter, it is not as expected, no new lines are added. Could you please help to confirm whether I'm using the correct settings for newline delimiter?

My steps as below:
I've set `"cos.object.record.delimiter": "\n"` in cos-sink json configuration,  the configuration as below, and the object uploaded to COS is in attachment:
[kafka-java-console-sample-topic-1-0000000025904443-0000000025904447.txt](https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/files/3528175/kafka-java-console-sample-topic-1-0000000025904443-0000000025904447.txt)


```
# curl http://localhost:8083/connectors/cos-sink-demo
{
    "name": "cos-sink-demo",
    "config": {
        "connector.class": "com.ibm.eventstreams.connect.cossink.COSSinkConnector",
        "cos.bucket.resiliency": "cross-region",
        "tasks.max": "1",
        "topics": "kafka-java-console-sample-topic",
        "cos.bucket.name": "****",
        "cos.api.key": "****",
        "cos.service.crn": "crn:v1:bluemix:public:cloud-object-storage:global:a/****:****::",
        "cos.object.record.delimiter": "\n",
        "name": "cos-sink-demo",
        "cos.endpoint.visibility": "public",
        "value.converter": "org.apache.kafka.connect.storage.StringConverter",
        "cos.bucket.location": "us",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "cos.object.records": "5"
    },
    "tasks": [
        {
            "connector": "cos-sink-demo",
            "task": 0
        }
    ],
    "type": "sink"
}
```

**Comment by dpittner** [comment]
I think the problem is with encoding a line break in a json with Java. As the only use case for now is to use a new line, I've adapted the code to be a boolean true/false that will include a new line directly or nothing (the default). I've also added a readme entry and adjusted the tests. the branch in my fork is updated. However I don't see the change. @lijs-cn can you give this another try? @mimaison should I open another PR or is there a way for me to update the code in this fork (e.g. pushing to my fork is pretty much the only thing I can do)

**Comment by dpittner** [comment]
@lijs-cn please try using:  "cos.object.record.delimiter.nl": true with the new code, that should work regardless of config file encoding challenges

**Comment by mimaison** [comment]
@dpittner If you push to your branch `dpittner:recordDelimiter` that will update this PR.

**Comment by lijs-cn** [comment]
@dpittner I haven't found your code changes in this PR, so I verified using your personal clone directly, while sets ` "cos.object.record.delimiter.nl": true` in the connector config, it can generate COS object as expected, newlines are added for each event! I've also verified using IBM SQL Query to query from cos objects, works perfectly!

**Comment by dpittner** [comment]
@mimaison I think something is odd, I did push to the branch in my fork, but no change here (e.g. that's also how @lijs-cn got the code). Likely I'm doing something wrong?

**Comment by mimaison** [comment]
@dpittner Not sure what happened. I can see the new commit in your branch but that did not update the PR. Also I was able to find your commit in this repo but the branch info is missing: https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/commit/ab9feec56a8857f199054211f06b062af5634595 :confused:

The best is probably to just open a new PR.

**Comment by mimaison** [comment]
Ping @dpittner

**Comment by dpittner** [comment]
@mimaison thanks for the ping, was on vacation (actually Bristol area ;-)) will open a new PR, thanks for following up!

**Comment by dpittner** [comment]
done https://github.com/ibm-messaging/kafka-connect-ibmcos-sink/pull/6

**Pull request closed without merging**